packages/react/src/components/AccountSettings/ChangePassword/ChangePassword.tsx

```diff
@@ -1,53 +1,163 @@
+  }, [validators]);
+
+  /**
+   * Note that formValues and other states are passed in as props so that
+   * it does not depend on whether or not those states have been updated yet
+   */
+  const validateNewPassword = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword } = formValues;
+
+      return runFieldValidators(
+        newPassword,
+        passwordValidators,
+        eventType,
+        blurredFields.newPassword
+      );
+    },
+    [passwordValidators]
+  );
+
+  const validateConfirmPassword = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword, confirmPassword } = formValues;
+
+      const confirmPasswordValidator = getConfirmPasswordValidator(newPassword);
+
+      return runFieldValidators(
+        confirmPassword,
+        [confirmPasswordValidator],
+        eventType,
+        blurredFields.confirmPassword
+      );
+    },
+    []
+  );
+
+  const runValidation = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ) => {
+      const newPasswordErrors = validateNewPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+      const confirmPasswordErrors = validateConfirmPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+
+      const newValidationError = {
+        newPassword: newPasswordErrors,
+        confirmPassword: confirmPasswordErrors,
+      };
+
+      // only re-render if errors have changed
+      if (!isEqual(validationError, newValidationError)) {
+        setValidationError(newValidationError);
+      }
+    },
+    [validateConfirmPassword, validateNewPassword, validationError]
+  );
 
   /** Translations */
   // TODO: add AccountSettingsTextUtil to collect these strings
   const currentPasswordLabel = translate('Current Password');
   const newPasswordLabel = translate('New Password');
+  const confirmPasswordLabel = translate('Confirm Password');
   const updatePasswordText = translate('Update password');
 
   /** Event Handlers */
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     event.preventDefault();
 
     const { name, value } = event.target;
-    setFormValues({ ...formValues, [name]: value });
+
+    const newFormValues = { ...formValues, [name]: value };
+    runValidation(newFormValues, blurredFields, 'change');
+
+    setFormValues(newFormValues);
+  };
+
+  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+    event.preventDefault();
+
+    const { name } = event.target;
+    const newBlurredFields = { ...blurredFields, [name]: true };
+    runValidation(formValues, newBlurredFields, 'blur');
+    setBlurredFields({ ...blurredFields, [name]: true });
```

Comment:
Do we need to call `setBlurredFields` if the field has already been blurred? Why can't we just directly pass `newBlurredFields` here?

Reply:
Made optimization, and directly passed `newBlurredFields` in [e187a92](https://github.com/aws-amplify/amplify-ui/pull/2900/commits/e187a926bfedb42e0c9b131643548a4f77f901b7)